### DISCUSSION TITLE
Storing secrets in Directus.

### DISCUSSION THREAD

**Discussion by DeanKamali** [directus/directus#10591] [Archive]
Happy Friday and Holidays, everyone!

I wonder if Directus has a built way to store secrets and has a method to retrieve them?
I think it's a useful feature instead of implementing your solution or using a third-party library or service like SSM to achieve that.

Thanks

**Comment by rijkvanzanten** [comment]
Happy holidays! 🎄 

There's currently no way to have a value be encrypted/decrypted from within Directus. We do have support for hashes though (through argon2), which allow you to save a value, and then later compare raw strings against the saved hash.

How would you envisions an "encrypted" field like this to work within the platform? Would Directus encrypt/decrypt the value on the fly (so the end user doesn't notice it), or would you have to provide a secret-string in order to decrypt the value?

**Reply by DeanKamali** [reply]
Thanks for getting back to me on this one. 

I would envision this to work as you suggested, to have Directus encrypt/decrypt value on the fly, with an optional encryption key to be specified as an environment variable. 

As for how values would be displayed in the UI, values would be masked **** and real value to be revealed upon clicking on the icon. 

![image](https://user-images.githubusercontent.com/1252959/147478455-f5af91ea-b895-4da4-bf8f-55deacad49fd.png)

**Reply by benhaynes** [reply]
I like it! If we use a symmetric encryption method, then Directus will need some key to use. This can be stored at the global level (ENV?), field level (is DB secure enough?), or the user(s) will have to remember it. We could also use asymmetric encryption if needed. I assume we don't want to use hashing for this interface, which is only good for "checks".

What algorithm would we use?
* AES (super secure, and good to include)
* RSA (asymmetric)
* Triple DES
* Blowfish (I think we use this for our native passwords)

**Reply by rijkvanzanten** [reply]
> Blowfish (I think we use this for our native passwords)

We don't use encryption for the native passwords (we use hashing). The hashing engine we use is argon2.

Funnily enough, the screenshot example would use _hashing_, not _encryption_ as you'd never want to be able to recreate the original raw password from the stored hash value 🙂

**Reply by benhaynes** [reply]
Well, yes, a password (in the example) would likely use a hash... but to support the "eye" toggle we need to be able to decrypt it too. I think there's a benefit to having Hash and Encryption support for data Fields.

**Reply by rijkvanzanten** [reply]
Absolutely, not saying there isn't a benefit to having it 👍🏻

**Reply by binaryben** [reply]
If I may 3c to the discussion. Storing the encryption key in the database would likely be pointless since once someone has access to the database, they'd also then have the key to unlock the data. Even having it stored in env is slightly concerning to me, and definitely not in a file. Memory is ideal, but how you get it in memory at start is the challenge.

Although old, I wonder [if this practice of splitting the key is still relevant?](https://stackoverflow.com/a/726060) Seems like it's the equivalent of salting a hash and storing the salt elsewhere. I do also support the suggestion of providing support for external services to manage the keys off of the server.

My preferences would be for Directus to [support both options 1 and 4 from this list](https://security.stackexchange.com/a/12334) whilst getting said key from an external service at startup if using option 4.

Also, I'd want an encrypted field to be able to have read/write permissions dedicated to that field. Just because a user has access to a collection shouldn't mean they have accesses to an encrypted field in it.

**Reply by rijkvanzanten** [reply]
> Memory is ideal, but how you get it in memory at start is the challenge.

If you'd generate it in memory without pulling it from anything, it would mean you'd lose access to your encrypted data every time you restarted the server or horizontally scale the instance. By doing it thru option 4 on your shared list where you manually type it in, you lose the ability to do any auto scaling or automatic instance recovery 🤔

**Reply by binaryben** [reply]
> If you'd generate it in memory without pulling it from anything ...

Correct. Storing it in memory still requires pulling it into memory from _somewhere_ persistent (i.e., a cloud based secret management service).

Ideally the secret management service allows the generation of access tokens used to retrieve the secret from a protected endpoint. Although this seems like it's just obfuscation by using one secret to mask another:

* secret managers have tools to quickly invalidate compromised access tokens without having to invalidate the actual secret
* tokens can be restricted to being accessed only from certain IPs or within VPNs (i.e., if the token was stored in the Directus settings table and the db was leaked somehow, that token would be useless outside of the servers IP address - this doesn't prevent against someone having physical access to the machine of course); and
* I think some tools work so that users (and bad actors) on the machine cannot see the secret even once it has been retrieved locally - but I'm not sure on the technical implementation of how that is achieved sorry.

I think it's best to pick some prominent secret management solutions and look at how they are implemented, then choose which ones the Directus team would be willing to support. Here is an [article detailing how to retrieve a secret from AWS Secret Manager using the Node SDK](https://medium.com/intelliconnect-engineering/how-to-store-and-retrieve-your-secret-keys-using-aws-secrets-manager-727db222b785) which could be a good place to start. Once retrieved, put it in memory and use it from there.

**Reply by rijkvanzanten** [reply]
That to me is where environment variables come in. AWS etc have tools in place to inject env vars from secrets manager etc. I don't think that's something we're looking to implement in the application layer itself

**Reply by binaryben** [reply]
> I don't think that's something we're looking to implement in the application layer itself

That's a shame then but I understand the team needs to prioritise what features they think are most useful. 

> Even having [encryption keys] stored in env is slightly concerning to me

If a machine is compromised, any process can read the environment variables with ease. My concern is if Directus provides encrypted fields (and I hope they do), people will start using them in situations where PCI Compliance and other similar issues (complying with local laws for storing personal health data) come into play. It needs to be that someone with remote access to the machine cannot decrypt the data if they were to take a copy of it.

Storing it in an env (regardless of the method the key ends up in the env - AWS secret management or otherwise) means they have the data dump and the key. 

I hate to say it, but I suspect it would be better for Directus to simply not provide an encrypted field than to implement it poorly.

**Comment by zhibirc** [comment]
@rijkvanzanten is it possible in some way to use AWS Secrets Manager (or other similar solution) for retrieving necessary environment variables (for example,  for STORAGE_S3_KEY and STORAGE_S3_SECRET)? Storing them as just clear text environment variables is not enough secure from my standpoint.

**Reply by rijkvanzanten** [reply]
Depends on how / where you're running Directus itself. Both ECS and EKS have ways to inject those from secrets manager on the fly. CloudFormation has a dedicated syntax to inject those variables into template parameters as well 👍🏻 

This feature request is about _using_ Directus as an alternative for Secrets Manager, not about being able to use Secrets Manager with Directus 🙂

**Reply by zhibirc** [reply]
Got the point! Many thanks for information about injecting secrets to ECS on the fly!

**Comment by maazamaani** [comment]
Hello there
I want to mention a use case for storing a hashed value in database that is decodabale in directus..

> We have implemented a multi tenant app in directus.. so we have a 'tenants' collection that is storing api keys for each tenant 3rd party service provider (like sms, payment gateway, etc.)

In this scenario, api keys cannot be stored directly in database. while cannot be stored in env because api keys are depend on tanant.. if we could store a hashed value in database and later decode using key-secret, is there any security concern? Is there any better solution?
